### PR TITLE
[TensorAlgebra] `BlockSparseArray` tensor contraction

### PR DESCRIPTION
**Pull request by mtfishman**
This is work in progress towards getting `BlockSparseArray` tensor contraction working (for now, independent of symmetries).

The design right now is that `BlockSparseArray` contraction is implemented through the generic matricized contract code in `TensorAlgebra`. This relies on having proper implementations of permutation, generalized reshape (which I call `fusedims` and `splitdims` in `TensorAlgebra`), and matrix multiplication, so the actual implementation of block sparse array contraction mostly relies on the generic one in `TensorAlgebra` and relies on those operations being defined.

The current matricized `contract` code already works for other array types like `DiagonalArray` and `SparseArrayDOK` since they implement the required permutation, reshape, and matrix multiplication functionality.

This will generalize to symmetric tensors with even more specalized implementations of `fusedims` and `splitdims` that actually fuse and split the blocks according to symmetry sector fusion, but that will be left for future work and will rely on the work in #1296 (though we could start testing it without that by making simple abelian group irrep types and using `GradedAxes` as the axes of a `BlockSparseArray`).

### CODE REVIEW

**Comment by emstoudenmire** [comment]
I'll be really interested to see how the performance turns out when using the generic matricized contract. It would be great if it's already quite good in many cases. What do you expect or do you already have some timings?

**Comment by mtfishman** [comment]
> I'll be really interested to see how the performance turns out when using the generic matricized contract. It would be great if it's already quite good in many cases. What do you expect or do you already have some timings?

Yeah I'm curious about that too. I think I could argue that if written efficiently, matricized contraction is either about the same as what we are doing currently if no QN merging is done (it requires the same tensor permutations of each block of the input tensors, and then once those permutations are done it requires the same matrix multiplication), and then could be faster if QN merging is done (depending on the particulars of the symmetries, sparsity pattern, tensor order, etc.). Currently I'm focusing more on code organization, generic interfaces, and correctness, but once the code structure is in place and working the next step would be to do benchmarking/profiling and optimization.

File: NDTensors/src/lib/BlockSparseArrays/src/blocksparsearrayinterface/blocksparsearrayinterface.jl [hunk 1]
```diff
@@ -45,10 +45,20 @@ function block_nstored(a::AbstractArray)
...
+
+## function blocksparse_copyto!(a_dest::AbstractArray, a_src::AbstractArray)
+##   blocksparse_map!(identity, a_dest, a_src)
+##   return a_dest
+## end
+## 
+## function blocksparse_permutedims!(a_dest, a_src::AbstractArray, perm)
+##   blocksparse_map!(identity, a_dest, PermutedDimsArray(a_src, perm))
+##   return a_dest
+## end
```
Comment (mtfishman) [inline NDTensors/src/lib/BlockSparseArrays/src/blocksparsearrayinterface/blocksparsearrayinterface.jl:61]:
```suggestion
```

File: NDTensors/src/lib/TensorAlgebra/src/fusedims.jl [hunk 2]
```diff
@@ -1,21 +1,5 @@
...
-  a_dest[] = a_src[]
-  return a_dest
-end
-function _permutedims(a::AbstractArray{<:Any,N}, perm::Tuple{Vararg{Int,N}}) where {N}
-  return permutedims(a, perm)
-end
-function _permutedims(a::AbstractArray{<:Any,0}, perm::Tuple{})
-  return copy(a)
-end
+# include("BaseExtensions/BaseExtensions.jl")
```
Comment (mtfishman) [inline NDTensors/src/lib/TensorAlgebra/src/fusedims.jl:1]:
```suggestion
```

File: NDTensors/src/lib/SparseArrayInterface/src/sparsearrayinterface/interface.jl [hunk 3]
```diff
@@ -3,8 +3,10 @@
 
 # Minimal interface
 # Data structure of the stored (generally nonzero) values
-# nonzeros(a::AbstractArray) = a
-sparse_storage(a::AbstractArray) = a
+# nonzeros(a::AbstractArray) = error("Not implemented")
```
Comment (mtfishman) [inline NDTensors/src/lib/SparseArrayInterface/src/sparsearrayinterface/interface.jl:6]:
```suggestion
```

**Pull request merged**